deploy-agent/deployd/client/client.py

```diff
@@ -29,31 +29,64 @@
 
 
 class Client(BaseClient):
-    def __init__(self, config=None, hostname=None,
-                 ip=None, hostgroup=None, host_id=None, use_facter=None):
+    def __init__(self, config=None, hostname=None, ip=None, hostgroup=None, 
+                host_id=None, use_facter=None, autoscaling_group=None, availability_zone=None):
         self._hostname = hostname
         self._ip = ip
         self._hostgroup = hostgroup
         self._id = host_id
         self._config = config
         self._use_facter = use_facter
+        self._agent_version = self._config.get_deploy_agent_version()
+        self._autoscaling_group = autoscaling_group
+        self._availability_zone = availability_zone
 
     def _read_host_info(self):
         if self._use_facter:
             log.info("Use facter to get host info")
-            if not self._hostname and self._config.get_facter_name_key():
-                self._hostname = utils.get_info_from_facter(self._config.get_facter_name_key())
+            name_key = self._config.get_facter_name_key()
+            ip_key = self._config.get_facter_ip_key()
+            id_key = self._config.get_facter_id_key()
+            group_key = self._config.get_facter_group_key()
+            az_key = self._config.get_facter_az_key()
+            keys_to_fetch = set()
+            # facter call is expensive so collect all keys to fetch first
```

Comment:
How expensive it is? is it just reading from cache in file? If it is related to the facter cache file lock, can we only call the facter once during the first deployd get_info_from_facter and store these info in a deployd property file? assuming these info will never change after the host launched.

Reply:
It's not related to cache file. For hostname, hostip, hostid, group, we currently exec "facter -p" 4 times(one for each). Jayme mentioned we can call "facter -p" with list of keys. So, converting those 4 calls into single call. 

Do we want to further put that in file and never call facter again?

Reply:
r+ now. we can iterate to improve later.